### PR TITLE
Fix bug in all-null list due to join_list_elements special handling

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
All-null list is considered as empty list in `join_list_elements`. So, nulls of children are replaced by `narep` before passing to `join_list_elements` API.
Related https://github.com/rapidsai/cudf/issues/12766

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/json/write_json.cu [hunk 1]
```diff
@@ -417,21 +417,28 @@ struct column_to_strings_fn {
...
+              children_names.size() > child_index ? children_names[child_index].children
+                                                            : std::vector<column_name_info>{});
+          } else {
+            return cudf::type_dispatcher(child_view.type(), *this, child_view);
+          }
+        }()
+          ->view(),
+        narep,
+        stream_,
+        mr_);
```
Comment (ttnghia) [inline cpp/src/io/json/write_json.cu:441]: This `child_string` is temporary variable thus we may use default `mr` for it.

Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu:441]:
```suggestion
        rmm::mr::get_current_device_resource());
```

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/json/write_json.cu [hunk 2]
```diff
@@ -417,21 +417,28 @@ struct column_to_strings_fn {
...
-            child_view,
-            children_names.size() > child_index ? children_names[child_index].children
-                                                          : std::vector<column_name_info>{});
-        } else {
-          return cudf::type_dispatcher(child_view.type(), *this, child_view);
-        }
-      }();
+      // nulls are replaced due to special handling of all-null lists as empty lists
+      // by join_list_elements
+      auto child_string = cudf::strings::detail::replace_nulls(
```
Comment (bdice) [inline cpp/src/io/json/write_json.cu:422]: Non-blocker: I think I'd prefer this to be two separate lines. First the IIFE, then call `cudf::strings::detail::replace_nulls`. It's just pretty long and complicated to read. After 20 lines of code, I forget that we're computing a function argument.

**Review by bdice — APPROVED** [review]

File: cpp/src/io/json/write_json.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu:422]: I used this to avoid the temporary. Updated.

File: cpp/src/io/json/write_json.cu [hunk 3]
```diff
@@ -431,7 +433,9 @@ struct column_to_strings_fn {
         } else {
           return cudf::type_dispatcher(child_view.type(), *this, child_view);
         }
-      }();
+      };
```
Comment (ttnghia) [inline cpp/src/io/json/write_json.cu:436]:
```suggestion
      }();
```

Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu:436]: It's written this way to avoid the temporary variable. It will help reduce peak memory.

**Pull request merged**